core/Enclave/key_factory.cpp

```diff
@@ -0,0 +1,570 @@
+                                     &(keyblob_data->payload[keyblob_data->ciphertext_size]), real_aad_size,
+                                     reinterpret_cast<uint8_t(*)[16]>(keyblob_data->mac));
+    if (SGX_SUCCESS != ret)
+    {
+        printf("gcm encrypting failed.\n");
+    }
+    else
+    {
+        keyblob_data->ciphertext_size = plaintext_size;
+        keyblob_data->aad_size = real_aad_size;
+    }
+
+    return ret;
+}
+
+// use the g_domain_key to decrypt the cmk and get it plaintext
+sgx_status_t ehsm_parse_keyblob(uint8_t *plaintext, uint32_t plaintext_size,
+                                const sgx_aes_gcm_data_ex_t *keyblob_data)
+{
+    if (NULL == keyblob_data || NULL == plaintext || NULL == plaintext_size || plaintext_size < keyblob_data->ciphertext_size)
+        return SGX_ERROR_INVALID_PARAMETER;
+
+    sgx_status_t ret = sgx_rijndael128GCM_decrypt(&g_domain_key,
+                                                  keyblob_data->payload, keyblob_data->ciphertext_size,
+                                                  plaintext,
+                                                  keyblob_data->iv, sizeof(keyblob_data->iv),
+                                                  &(keyblob_data->payload[keyblob_data->ciphertext_size]), keyblob_data->aad_size,
+                                                  (const sgx_aes_gcm_128bit_tag_t *)keyblob_data->mac);
+    if (SGX_SUCCESS != ret)
+        printf("gcm decrypting failed.\n");
+    else
+        plaintext_size = keyblob_data->ciphertext_size;
+
+    return ret;
+}
+
+uint32_t ehsm_get_symmetric_key_size(ehsm_keyspec_t key_spec)
+{
+    switch (key_spec)
+    {
+    case EH_AES_GCM_128:
+    case EH_SM4_CTR:
+    case EH_SM4_CBC:
+        return 16;
+    case EH_AES_GCM_192:
+        return 24;
+    case EH_AES_GCM_256:
+        return 32;
+    default:
+        return 0;
+    }
+    return 0;
+}
+
+/**
+ * @brief generate aes_gcm key with openssl api
+ * @param cmk_blob storage key information
+ * @param APPEND_SIZE_TO_KEYBOB_T the size of cmk_blob
+ * @param req_blob_size the gcm data size
+ * @param keyspec key type, refer to [ehsm_keyspec_t]
+ * @return sgx_status_t
+ */
+sgx_status_t ehsm_create_aes_key(ehsm_keyblob_t *cmk)
+{
+    sgx_status_t ret = SGX_ERROR_UNEXPECTED;
+
+    if (cmk->metadata.keyspec != EH_AES_GCM_128 &&
+        cmk->metadata.keyspec != EH_AES_GCM_192 &&
+        cmk->metadata.keyspec != EH_AES_GCM_256)
+    {
+        return SGX_ERROR_INVALID_PARAMETER;
+    }
+
+    uint32_t keysize = ehsm_get_symmetric_key_size(cmk->metadata.keyspec);
+    if (keysize == 0)
+    {
+        return SGX_ERROR_UNEXPECTED;
+    }
+    uint32_t real_blob_size = ehsm_calc_keyblob_size(cmk->metadata.keyspec);
+
+    if (real_blob_size == UINT32_MAX)
+    {
+        return SGX_ERROR_UNEXPECTED;
+    }
+    if (cmk->keybloblen == 0)
+    {
+        cmk->keybloblen = real_blob_size;
+        return SGX_SUCCESS;
+    }
+
```

Comment:
for each entry, you can change the logic as  (combine the line206-line217):
if (cmk->keybloblen == 0) {
     cmk->keybloblen = ehsm_calc_keyblob_size(cmk->metadata.keyspec);
     return SGX_SUCCESS;
}

// handle the generic logic to create the key.